// anujs/aaa/src/npm/schnee-ui/components/XNavigator/index.js
import React from "../../../../ReactQuick.js";

const noop = function() {};

function XNavigator(props) {
    if (typeof props.children !== 'string') {
        console.warn('<Navigator>组件的内容暂时只能传字符串，不能放标签');
    }

    this.state = {
        children: props.children
    };
}

XNavigator = React.toClass(XNavigator, React.Component, {
    componentWillReceiveProps: function(props) {
        if (this.state.children != props.children) {
            this.setState({
                children: props.children
            });
        }
    },
    goPage: function() {
        var method = this.props.openType;
        var hook = methodMap[method] || 'navigateTo';
        React.api[hook]({
            url: this.props.url,
            success: this.props.bindsuccess || noop,
            fail: this.props.bindfail || noop,
            complete: this.props.bindcomplete || noop
        });
    },
    render: function() {
        var h = React.createElement;
        return h("div", {
            class: 'anu-navigator ' + this.props.class,
            onClick: this.goPage.bind(this),
            "data-click-uid": "e39_62",
            "data-beacon-uid": "default"
        }, this.state.children);;
    },
    classUid: "c1306"
}, {});
XNavigator.defaultProps = {
    target: 'self',
    url: '',
    openType: 'navigate',
    hoverClass: 'navigator-hover'
};
const methodMap = {
    navigate: 'navigateTo',
    redirect: 'redirectTo',
    switchTab: 'switchTab',
    reLaunch: 'reLaunch',
    navigateBack: 'navigateBack'
};
export {
    React
};
export default XNavigator;
