// anujs/aaa/src/npm/schnee-ui/components/XOverlay/index.js
import React from "../../../../ReactQuick.js";

function XOverlay(props) {
    this.state = {
        show: false,
        className: ''
    };
    this.overlayClick = this.overlayClick.bind(this);
    this.handleTouchMove = this.handleTouchMove.bind(this);
}

XOverlay = React.toClass(XOverlay, React.Component, {
    componentDidMount: function() {
        this.updateVisible(this.props.visible);
    },
    componentWillReceiveProps: function(nextProps) {
        if (this.props.visible !== nextProps.visible) {
            this.updateVisible(nextProps.visible);
        }
    },
    shouldComponentUpdate: function(nextProps, nextState) {
        if (nextProps.visible === this.props.visible && nextState.show === this.state.show && nextState.className === this.state.className) {
            return false;
        }

        return true;
    },
    updateVisible: function(visible) {
        this.timeoutId && clearTimeout(this.timeoutId);

        if (visible) {
            this.setState({
                show: true,
                className: 'quist-overlay-enter'
            });
        } else {
            this.setState({
                className: 'quist-overlay-leave'
            });
            this.timeoutId = setTimeout(() => {
                this.setState({
                    show: false
                });
            }, 500);
        }
    },
    overlayClick: function() {
        this.setState({
            show: false
        });
        this.props.onClose && this.props.onClose();
    },
    handleTouchMove: function(e) {
        return false;
    },
    render: function() {
        var h = React.createElement;
        return h("div", {
            className: 'quist-overlay-mask  ' + this.state.className,
            onClick: this.overlayClick,
            catchTouchMove: this.handleTouchMove,
            style: React.toStyle({
                backgroundColor: this.props.background,
                display: this.state.show ? 'flex' : 'none'
            }, this.props, 'style3278'),
            "data-click-uid": "e77_10",
            "data-beacon-uid": "default",
            "data-touchmove-uid": "e78_10"
        });;
    },
    classUid: "c1874"
}, {});
XOverlay.defaultProps = {
    maskClosable: false,
    visible: false,
    background: 'rgba(0, 0, 0, 0.6)'
};
export {
    React
};
export default XOverlay;
